relations/src/shielder/merge.rs

```diff
@@ -0,0 +1,319 @@
+        let first_old_nullifier = FpVar::new_input(ns!(cs, "first old nullifier"), || {
+            self.first_old_nullifier()
+        })?;
+        let first_old_note =
+            FpVar::new_witness(ns!(cs, "first old note"), || self.first_old_note())?;
+
+        check_note(
+            &token_id,
+            &first_old_token_amount,
+            &first_old_trapdoor,
+            &first_old_nullifier,
+            &first_old_note,
+        )?;
+
+        //------------------------------
+        // Check second old note arguments.
+        //------------------------------
+        let second_old_token_amount =
+            FpVar::new_witness(ns!(cs, "second old token amount"), || {
+                self.second_old_token_amount()
+            })?;
+        let second_old_trapdoor = FpVar::new_witness(ns!(cs, "second old trapdoor"), || {
+            self.second_old_trapdoor()
+        })?;
+        let second_old_nullifier = FpVar::new_input(ns!(cs, "second old nullifier"), || {
+            self.second_old_nullifier()
+        })?;
+        let second_old_note =
+            FpVar::new_witness(ns!(cs, "second old note"), || self.second_old_note())?;
+
+        check_note(
+            &token_id,
+            &second_old_token_amount,
+            &second_old_trapdoor,
+            &second_old_nullifier,
+            &second_old_note,
+        )?;
+
+        //------------------------------
+        // Check new note arguments.
+        //------------------------------
+        let new_token_amount =
+            FpVar::new_witness(ns!(cs, "new token amount"), || self.new_token_amount())?;
+        let new_trapdoor = FpVar::new_witness(ns!(cs, "new trapdoor"), || self.new_trapdoor())?;
+        let new_nullifier = FpVar::new_witness(ns!(cs, "new nullifier"), || self.new_nullifier())?;
+        let new_note = FpVar::new_input(ns!(cs, "new note"), || self.new_note())?;
+
+        check_note(
+            &token_id,
+            &new_token_amount,
+            &new_trapdoor,
+            &new_nullifier,
+            &new_note,
+        )?;
+
+        //----------------------------------
+        // Check token value soundness.
+        //----------------------------------
+        // some range checks for overflows?
+        let token_sum = first_old_token_amount.add(second_old_token_amount);
+        token_sum.enforce_equal(&new_token_amount)?;
+
+        //------------------------
+        // Check first merkle proof.
+        //------------------------
+        let merkle_root = FpVar::new_input(ns!(cs, "merkle root"), || self.merkle_root())?;
+        let first_path_shape = PathShapeVar::new_witness(ns!(cs, "first path shape"), || {
+            Ok((*self.max_path_len(), self.first_leaf_index().cloned()))
+        })?;
+
+        check_merkle_proof(
+            merkle_root.clone(),
+            first_path_shape,
+            first_old_note,
+            self.first_merkle_path().cloned().unwrap_or_default(),
```

Comment:
maybe let's use `unwrap()` - the path should never be empty, and if it is, we should panic

Reply:
This actually not true in our tests so far ;) Key generation uses an empty circuit for which the constraints are generated here https://github.com/Cardinal-Cryptography/aleph-node/blob/0af5afb4356829d56425e2dac16c7a7253141762/relations/src/shielder/deposit_and_merge.rs#L221. This is the likely reason why @mike1729 went for `unwrap_or_default`. The way I see it, we can leave the default version and add additional checks in the tests. I am open to other suggestions, though.